rust/extns/numaflow-kafka/src/lib.rs

```diff
@@ -0,0 +1,680 @@
+            } => {
+                let status = self.ack_messages(offsets).await;
+                let _ = respond_to.send(status);
+            }
+            KafkaActorMessage::Pending { respond_to } => {
+                let pending = self.pending_messages().await;
+                let _ = respond_to.send(pending);
+            }
+        }
+    }
+
+    async fn read_messages(&mut self) -> Result<Vec<KafkaMessage>> {
+        let mut messages: Vec<KafkaMessage> = vec![];
+        let timeout = tokio::time::timeout(self.read_timeout, std::future::pending::<()>());
+        tokio::pin!(timeout);
+        loop {
+            if messages.len() >= self.batch_size {
+                break;
+            }
+            tokio::select! {
+                biased;
+
+                _ = &mut timeout => {
+                    break;
+                }
+
+                message = self.consumer.recv() => {
+                    let message = match message {
+                        Ok(msg) => msg,
+                        Err(e) => {
+                            error!(?e, "Failed to read messages, will retry after 1 second");
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            continue;
+                        }
+                    };
+
+                    let headers = match message.headers() {
+                        Some(headers) => headers
+                            .iter()
+                            .map(|header| {
+                                (
+                                    header.key.to_string(),
+                                    String::from_utf8_lossy(header.value.unwrap_or_default()).to_string(),
+                                )
+                            })
+                            .collect(),
+                        None => HashMap::new(),
+                    };
+
+                    let value = match message.payload() {
+                        Some(payload) => Bytes::copy_from_slice(payload),
+                        None => Bytes::new(),
+                    };
+
+                    let message = KafkaMessage {
+                        value,
+                        partition: message.partition(),
+                        offset: message.offset(),
+                        headers,
+                    };
+
+                    messages.push(message);
+                }
+            }
+        }
+        tracing::debug!(msg_count = messages.len(), "Read messages from Kafka");
+        Ok(messages)
+    }
+
+    async fn ack_messages(&mut self, offsets: Vec<(i32, i64)>) -> Result<()> {
+        let mut tpl = TopicPartitionList::new();
+        for (partition, offset) in offsets {
+            tpl.add_partition_offset(&self.topic, partition, Offset::Offset(offset + 1))
+                .map_err(|e| Error::Kafka(format!("Failed to add partition offset: {}", e)))?;
+        }
+
+        self.consumer
+            .commit(&tpl, CommitMode::Async)
```

Comment:
should we support different modes? Async will end up in replays.

Reply:
Changed to `CommitMode::Sync`